atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+                prep_file = PreparedFile.from_config(self.orig_file, client=client)
+                asyncio.run(prep_file.fill_cache())
+            elif isinstance(self.orig_file, ProcedureFile):
+                prep_file = PreparedProcedureFile.from_origin(self.orig_file)
+
+            cache = get_signal_cache()
+            self._signals = list(cache.keys())
+            self._devices = list(happi.loader.cache.keys())
+
+    def _fill_devices_list(self) -> None:
+        self.device_table.setRowCount(max(len(self._signals), len(self._devices)))
+        for i, sig in enumerate(self._signals):
+            self.device_table.setItem(i, 1, QtWidgets.QTableWidgetItem(sig))
+        for i, dev in enumerate(self._devices):
+            self.device_table.setItem(i, 0, QtWidgets.QTableWidgetItem(dev))
+
+    def verify_changes(self) -> None:
+        verify_changes(self, self.orig_file)
+
+    def save_file(self) -> None:
+        # open save message box
+        self.prompt_apply()
+        filename, _ = QtWidgets.QFileDialog.getSaveFileName(
+            parent=self,
+            caption='Save as',
+            filter='Json Files (*.json)',
+        )
+        if not filename.endswith('.json'):
+            filename += '.json'
+
+        # get serialized
+        serialized = serialize(type(self.orig_file), self.orig_file)
+        try:
+            with open(filename, 'w') as fd:
+                json.dump(serialized, fd, indent=2)
+                # Ends file on newline as per pre-commit
+                fd.write('\n')
+        except OSError:
+            logger.exception(f'Error saving file {filename}')
+
+        # inform about saving
+        reply = QtWidgets.QMessageBox.question(
+            self,
+            'File saved',
+            'File saved successfully, would you like to open this file?'
+        )
+
+        if reply == QtWidgets.QMessageBox.Yes:
+            self._window._new_tab(data=self.orig_file, filename=filename)
+
+    def apply_all(self) -> None:
+        self.prompt_apply()
+        self.update_title()
+
+    def prompt_apply(self) -> None:
+        # message box with details on remaining changes
+        self.update_change_list()
+        if len(self.all_actions) <= 0:
+            return
+
+        reply = QtWidgets.QMessageBox.question(
+            self,
+            'Apply remaning edits?',
+            (
+                'Would you like to apply the remaining '
+                f'({len(self.all_actions)}) edits?'
+            )
+        )
+        if reply == QtWidgets.QMessageBox.Yes:
+            for action in self.all_actions:
+                action.apply()
+
+            # clear all rows
+            self.edits_table.clearContents()
+            self.details_list.clear()
+
+    def update_title(self) -> None:
+        """
+        Update the title.  Will be the name and the number of unapplied edits
+        """
+        file_name = os.path.basename(self.fp)
```

Comment:
Dealing with `None` is annoying with respect to the static analyzer but I find it to be good practice - even if you're 100% sure this doesn't get called when `fp is None` - to guard for it anyway. A future refactor you may forget about it and then get confused why `basename(None)` is getting called